File: 0-99/3-longest-substring-without-repeating-characters.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	var (
		start, end = 0, 0
		max        int
		freq       = make(map[byte]bool)
	)
	for end < len(s) {
		// Start move forward
		for start <= end && freq[s[end]] {
			freq[s[start]] = false
			start++
		}
		// End move forward
		for end < len(s)-1 {
			freq[s[end]] = true
			if freq[s[end+1]] {
				break
			}
			end++
		}
		length := end - start + 1
		if length > max {
			max = length
		}
		end++
	}
	return max
}

func main() {
	fmt.Println(lengthOfLongestSubstring("bbbbb"))
}

File: 0-99/3-longest-substring-without-repeating-characters.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLength = 0
        freqMap = defaultdict(int)
        i, j = 0, 0
        
        while j < len(s):
            # move j forward
            while j < len(s) and freqMap[s[j]] < 1:
                freqMap[s[j]] += 1
                j += 1
                
            maxLength = max(maxLength, j - i)
            # move i forward
            print(i, j, freqMap)
            while j < len(s) and i < j and freqMap[s[j]] >= 1:
                freqMap[s[i]] -= 1
                i += 1
      
        return maxLength


s = Solution()
print(s.lengthOfLongestSubstring("abcabcbb"))
